File: apps/backend/src/workspaces_tulearn_home/access.rs
use super::*;

// ---------------------------------------------------------------------------
// Subject resolution (port of access.ts::resolveTulearnSubject).
// ---------------------------------------------------------------------------

pub(super) enum TulearnError {
    /// Maps to TulearnAccessError (403/404).
    Access { status: u16, message: &'static str },
    /// Any unexpected failure -> 500.
    Internal,
}

pub(super) struct TulearnSubject {
    pub(super) role: &'static str,
    pub(super) read_only: bool,
    pub(super) ws_id: String,
    pub(super) student_platform_user_id: String,
    pub(super) student_workspace_user_id: String,
    pub(super) student_name: Option<String>,
}

pub(super) async fn resolve_tulearn_subject(
    contact_data: &contact::ContactDataConfig,
    outbound: &impl OutboundHttpClient,
    raw_ws_id: &str,
    user_id: &str,
    access_token: &str,
    student_id: Option<&str>,
) -> Result<TulearnSubject, TulearnError> {
    let ws_id = normalize_workspace_id(contact_data, outbound, raw_ws_id, user_id, access_token)
        .await
        .map_err(|()| TulearnError::Internal)?;

    if !has_education_enabled(contact_data, outbound, &ws_id)
        .await
        .map_err(|()| TulearnError::Internal)?
    {
        return Err(TulearnError::Access {
            status: 404,
            message: TULEARN_DISABLED_MESSAGE,
        });
    }

    let self_student = resolve_student_for_platform_user(contact_data, outbound, &ws_id, user_id)
        .await
        .map_err(|()| TulearnError::Internal)?;

    if student_id.is_none()
        && let Some(self_student) = self_student
    {
        return Ok(TulearnSubject {
            role: "student",
            read_only: false,
            ws_id,
            student_platform_user_id: user_id.to_owned(),
            student_workspace_user_id: self_student.workspace_user_id,
            student_name: self_student.name,
        });
    }

    // Parent link lookup.
    let link = parent_student_link(contact_data, outbound, &ws_id, user_id, student_id)
        .await
        .map_err(|()| TulearnError::Internal)?;
    let Some(link) = link else {
        return Err(TulearnError::Access {
            status: 403,
            message: NO_LEARNER_ACCESS_MESSAGE,
        });
    };

    let student_name = workspace_user_display_name(
        contact_data,
        outbound,
        &ws_id,
        &link.student_workspace_user_id,
    )
    .await
    .map_err(|()| TulearnError::Internal)?;

    Ok(TulearnSubject {
        role: "parent",
        read_only: true,
        ws_id,
        student_platform_user_id: link.student_platform_user_id,
        student_workspace_user_id: link.student_workspace_user_id,
        student_name,
    })
}

struct SelfStudent {
    workspace_user_id: String,
    name: Option<String>,
}

#[derive(Deserialize)]
struct WorkspaceUserRow {
    full_name: Option<String>,
    display_name: Option<String>,
    email: Option<String>,
}

fn to_display_name(
    display_name: Option<&str>,
    full_name: Option<&str>,
    email: Option<&str>,
) -> Option<String> {
    // Mirrors helpers.ts::toDisplayName: display_name || full_name || email || null.
    [display_name, full_name, email]
        .into_iter()
        .flatten()
        .map(str::trim)
        .find(|value| !value.is_empty())
        .map(ToOwned::to_owned)
}

#[derive(Deserialize)]
struct LinkedUserRow {
    virtual_user_id: Option<String>,
    workspace_users: Option<Value>,
}

async fn resolve_student_for_platform_user(
    contact_data: &contact::ContactDataConfig,
    outbound: &impl OutboundHttpClient,
    ws_id: &str,
    platform_user_id: &str,
) -> Result<Option<SelfStudent>, ()> {
    let Some(url) = contact_data.rest_url(
        "workspace_user_linked_users",
        &[
            (
                "select",
                "virtual_user_id,workspace_users!inner(id,full_name,display_name,email,avatar_url,ws_id)"
                    .to_owned(),
            ),
            ("platform_user_id", format!("eq.{platform_user_id}")),
            ("ws_id", format!("eq.{ws_id}")),
            ("limit", "1".to_owned()),
        ],
    ) else {
        return Err(());
    };
    let response = service_role_get(contact_data, outbound, &url, None).await?;
    if !(200..300).contains(&response.status) {
        return Err(());
    }

    let rows: Vec<LinkedUserRow> = response.json().map_err(|_| ())?;
    let Some(row) = rows.into_iter().next() else {
        return Ok(None);
    };

    let Some(virtual_user_id) = row.virtual_user_id.filter(|id| !id.trim().is_empty()) else {
        return Ok(None);
    };
    // PostgREST may return an embedded one-to-one as object or single-element array.
    let workspace_user = first_object(row.workspace_users.as_ref());
    let Some(workspace_user) = workspace_user else {
        return Ok(None);
    };

    let name = to_display_name(
        workspace_user.get("display_name").and_then(Value::as_str),
        workspace_user.get("full_name").and_then(Value::as_str),
        workspace_user.get("email").and_then(Value::as_str),
    );

    Ok(Some(SelfStudent {
        workspace_user_id: virtual_user_id,
        name,
    }))
}

struct ParentLink {
    student_platform_user_id: String,
    student_workspace_user_id: String,
}

#[derive(Deserialize)]
struct ParentLinkRow {
    student_platform_user_id: Option<String>,
    student_workspace_user_id: Option<String>,
}

async fn parent_student_link(
    contact_data: &contact::ContactDataConfig,
    outbound: &impl OutboundHttpClient,
    ws_id: &str,
    parent_user_id: &str,
    student_id: Option<&str>,
) -> Result<Option<ParentLink>, ()> {
    let mut params = vec![
        (
            "select",
            "student_platform_user_id,student_workspace_user_id".to_owned(),
        ),
        ("ws_id", format!("eq.{ws_id}")),
        ("parent_user_id", format!("eq.{parent_user_id}")),
        ("status", "eq.active".to_owned()),
    ];
    if let Some(student_id) = student_id {
        params.push(("student_workspace_user_id", format!("eq.{student_id}")));
    }
    params.push(("order", "created_at.asc".to_owned()));
    params.push(("limit", "1".to_owned()));

    let Some(url) = contact_data.rest_url("tulearn_parent_student_links", &params) else {
        return Err(());
    };
    let response = service_role_get(contact_data, outbound, &url, None).await?;
    if !(200..300).contains(&response.status) {
        return Err(());
    }

    let rows: Vec<ParentLinkRow> = response.json().map_err(|_| ())?;
    Ok(rows.into_iter().next().and_then(|row| {
        match (row.student_platform_user_id, row.student_workspace_user_id) {
            (Some(platform), Some(workspace)) => Some(ParentLink {
                student_platform_user_id: platform,
                student_workspace_user_id: workspace,
            }),
            _ => None,
        }
    }))
}

async fn workspace_user_display_name(
    contact_data: &contact::ContactDataConfig,
    outbound: &impl OutboundHttpClient,
    ws_id: &str,
    workspace_user_id: &str,
) -> Result<Option<String>, ()> {
    let Some(url) = contact_data.rest_url(
        "workspace_users",
        &[
            (
                "select",
                "id,full_name,display_name,email,avatar_url".to_owned(),
            ),
            ("id", format!("eq.{workspace_user_id}")),
            ("ws_id", format!("eq.{ws_id}")),
            ("limit", "1".to_owned()),
        ],
    ) else {
        return Err(());
    };
    let response = service_role_get(contact_data, outbound, &url, None).await?;
    if !(200..300).contains(&response.status) {
        return Err(());
    }

    let rows: Vec<WorkspaceUserRow> = response.json().map_err(|_| ())?;
    Ok(rows.into_iter().next().and_then(|row| {
        to_display_name(
            row.display_name.as_deref(),
            row.full_name.as_deref(),
            row.email.as_deref(),
        )
    }))
}

async fn has_education_enabled(
    contact_data: &contact::ContactDataConfig,
    outbound: &impl OutboundHttpClient,
    ws_id: &str,
) -> Result<bool, ()> {
    let Some(url) = contact_data.rest_url(
        "workspace_secrets",
        &[
            ("select", "value".to_owned()),
            ("ws_id", format!("eq.{ws_id}")),
            ("name", format!("eq.{ENABLE_EDUCATION_SECRET}")),
            ("limit", "1".to_owned()),
        ],
    ) else {
        return Err(());
    };
    let response = service_role_get(contact_data, outbound, &url, None).await?;
    if !(200..300).contains(&response.status) {
        return Err(());
    }

    #[derive(Deserialize)]
    struct SecretRow {
        value: Option<String>,
    }
    let rows: Vec<SecretRow> = response.json().map_err(|_| ())?;
    Ok(rows
        .into_iter()
        .next()
        .and_then(|row| row.value)
        .map(|value| value.trim().to_lowercase() == "true")
        .unwrap_or(false))
}

File: packages/education-core/src/tulearn/access.ts
import { normalizeWorkspaceId } from '@tuturuuu/utils/workspace-helper';
import { NextResponse } from 'next/server';

import { ENABLE_EDUCATION_SECRET } from './constants';
import { getAdmin } from './db';
import { toDisplayName } from './helpers';
import type {
  Db,
  ResolveTulearnSubjectInput,
  TulearnBootstrapInput,
  TulearnStudentSummary,
  TulearnSubject,
  TulearnWorkspaceSummary,
} from './types';

export class TulearnAccessError extends Error {
  constructor(
    message: string,
    public readonly status: 403 | 404
  ) {
    super(message);
    this.name = 'TulearnAccessError';
  }
}

export function tulearnAccessErrorResponse(error: unknown) {
  if (!(error instanceof TulearnAccessError)) return null;
  return NextResponse.json(
    { message: error.message },
    { status: error.status }
  );
}

export async function hasEducationEnabled(wsId: string, db?: Db) {
  const sbAdmin = await getAdmin(db);
  const { data, error } = await sbAdmin
    .from('workspace_secrets')
    .select('value')
    .eq('ws_id', wsId)
    .eq('name', ENABLE_EDUCATION_SECRET)
    .maybeSingle();

  if (error) throw error;
  return data?.value?.trim().toLowerCase() === 'true';
}

export async function resolveStudentForPlatformUser({
  db,
  platformUserId,
  wsId,
}: {
  db?: Db;
  platformUserId: string;
  wsId: string;
}) {
  const sbAdmin = await getAdmin(db);
  const { data, error } = await sbAdmin
    .from('workspace_user_linked_users')
    .select(
      'virtual_user_id, workspace_users!inner(id, full_name, display_name, email, avatar_url, ws_id)'
    )
    .eq('platform_user_id', platformUserId)
    .eq('ws_id', wsId)
    .maybeSingle();

  if (error) throw error;
  const workspaceUser = Array.isArray(data?.workspace_users)
    ? data?.workspace_users[0]
    : data?.workspace_users;

  if (!data?.virtual_user_id || !workspaceUser) return null;

  return {
    id: data.virtual_user_id,
    platform_user_id: platformUserId,
    workspace_user_id: data.virtual_user_id,
    workspace_id: wsId,
    name: toDisplayName(workspaceUser),
    email: workspaceUser.email ?? null,
    avatar_url: workspaceUser.avatar_url ?? null,
  } satisfies TulearnStudentSummary;
}

export async function resolveTulearnSubject({
  requestSupabase,
  studentId,
  user,
  wsId,
}: ResolveTulearnSubjectInput): Promise<TulearnSubject> {
  const normalizedWsId = await normalizeWorkspaceId(wsId, requestSupabase);
  const sbAdmin = await getAdmin();

  if (!(await hasEducationEnabled(normalizedWsId, sbAdmin))) {
    throw new TulearnAccessError(
      'Tulearn is not enabled for this workspace',
      404
    );
  }

  const selfStudent = await resolveStudentForPlatformUser({
    db: sbAdmin,
    platformUserId: user.id,
    wsId: normalizedWsId,
  });

  if (!studentId && selfStudent) {
    return {
      role: 'student',
      readOnly: false,
      wsId: normalizedWsId,
      studentPlatformUserId: user.id,
      studentWorkspaceUserId: selfStudent.workspace_user_id,
      studentName: selfStudent.name,
    };
  }

  let parentLinkQuery = sbAdmin
    .from('tulearn_parent_student_links')
    .select('student_platform_user_id, student_workspace_user_id')
    .eq('ws_id', normalizedWsId)
    .eq('parent_user_id', user.id)
    .eq('status', 'active');

  if (studentId) {
    parentLinkQuery = parentLinkQuery.eq(
      'student_workspace_user_id',
      studentId
    );
  }

  const { data: link, error } = await parentLinkQuery
    .order('created_at', { ascending: true })
    .limit(1)
    .maybeSingle();

  if (error) throw error;
  if (!link) {
    throw new TulearnAccessError("You don't have access to this learner", 403);
  }

  const { data: workspaceUser, error: workspaceUserError } = await sbAdmin
    .from('workspace_users')
    .select('id, full_name, display_name, email, avatar_url')
    .eq('id', link.student_workspace_user_id)
    .eq('ws_id', normalizedWsId)
    .maybeSingle();

  if (workspaceUserError) throw workspaceUserError;

  return {
    role: 'parent',
    readOnly: true,
    wsId: normalizedWsId,
    studentPlatformUserId: link.student_platform_user_id,
    studentWorkspaceUserId: link.student_workspace_user_id,
    studentName: workspaceUser ? toDisplayName(workspaceUser) : null,
  };
}

export async function getTulearnBootstrap({
  requestSupabase,
  user,
}: TulearnBootstrapInput) {
  const sbAdmin = await getAdmin();
  const [
    membershipWorkspacesResult,
    linkedUsersResult,
    parentLinksResult,
    profileResult,
    privateDetailsResult,
  ] = await Promise.all([
    sbAdmin
      .from('workspaces')
      .select(
        'id, name, avatar_url, logo_url, workspace_members!inner(user_id), workspace_secrets!inner(name, value)'
      )
      .eq('workspace_members.user_id', user.id)
      .eq('workspace_secrets.name', ENABLE_EDUCATION_SECRET)
      .eq('workspace_secrets.value', 'true'),
    sbAdmin
      .from('workspace_user_linked_users')
      .select('ws_id, virtual_user_id')
      .eq('platform_user_id', user.id),
    sbAdmin
      .from('tulearn_parent_student_links')
      .select('ws_id, student_platform_user_id, student_workspace_user_id')
      .eq('parent_user_id', user.id)
      .eq('status', 'active'),
    requestSupabase
      .from('users')
      .select('id, display_name, avatar_url')
      .eq('id', user.id)
      .maybeSingle(),
    requestSupabase
      .from('user_private_details')
      .select('email, full_name')
      .eq('user_id', user.id)
      .maybeSingle(),
  ]);

  if (membershipWorkspacesResult.error) throw membershipWorkspacesResult.error;
  if (linkedUsersResult.error) throw linkedUsersResult.error;
  if (parentLinksResult.error) throw parentLinksResult.error;
  if (profileResult.error) throw profileResult.error;
  if (privateDetailsResult.error) throw privateDetailsResult.error;

  const workspaceMap = new Map<string, TulearnWorkspaceSummary>();
  for (const workspace of membershipWorkspacesResult.data ?? []) {
    workspaceMap.set(workspace.id, {
      id: workspace.id,
      name: workspace.name ?? null,
      avatar_url: workspace.avatar_url ?? null,
      logo_url: workspace.logo_url ?? null,
      roles: ['student'],
    });
  }

  const linkedWorkspaceIds = [
    ...new Set((linkedUsersResult.data ?? []).map((link) => link.ws_id)),
  ];
  const linkedWorkspacesResult = linkedWorkspaceIds.length
    ? await sbAdmin
        .from('workspaces')
        .select(
          'id, name, avatar_url, logo_url, workspace_secrets!inner(name, value)'
        )
        .in('id', linkedWorkspaceIds)
        .eq('workspace_secrets.name', ENABLE_EDUCATION_SECRET)
        .eq('workspace_secrets.value', 'true')
    : { data: [], error: null };

  if (linkedWorkspacesResult.error) throw linkedWorkspacesResult.error;

  for (const workspace of linkedWorkspacesResult.data ?? []) {
    const existing = workspaceMap.get(workspace.id);
    if (existing) {
      if (!existing.roles.includes('student')) existing.roles.push('student');
      continue;
    }

    workspaceMap.set(workspace.id, {
      id: workspace.id,
      name: workspace.name ?? null,
      avatar_url: workspace.avatar_url ?? null,
      logo_url: workspace.logo_url ?? null,
      roles: ['student'],
    });
  }

  const parentLinks = parentLinksResult.data ?? [];
  const parentWorkspaceIds = [
    ...new Set(parentLinks.map((link) => link.ws_id)),
  ];
  const [parentWorkspacesResult, parentStudentsResult] = await Promise.all([
    parentWorkspaceIds.length
      ? sbAdmin
          .from('workspaces')
          .select(
            'id, name, avatar_url, logo_url, workspace_secrets!inner(name, value)'
          )
          .in('id', parentWorkspaceIds)
          .eq('workspace_secrets.name', ENABLE_EDUCATION_SECRET)
          .eq('workspace_secrets.value', 'true')
      : Promise.resolve({ data: [], error: null }),
    parentLinks.length
      ? sbAdmin
          .from('workspace_users')
          .select('id, ws_id, full_name, display_name, email, avatar_url')
          .in(
            'id',
            parentLinks.map((link) => link.student_workspace_user_id)
          )
      : Promise.resolve({ data: [], error: null }),
  ]);

  if (parentWorkspacesResult.error) throw parentWorkspacesResult.error;
  if (parentStudentsResult.error) throw parentStudentsResult.error;

  for (const workspace of parentWorkspacesResult.data ?? []) {
    const existing = workspaceMap.get(workspace.id);
    if (existing) {
      if (!existing.roles.includes('parent')) existing.roles.push('parent');
      continue;
    }
    workspaceMap.set(workspace.id, {
      id: workspace.id,
      name: workspace.name ?? null,
      avatar_url: workspace.avatar_url ?? null,
      logo_url: workspace.logo_url ?? null,
      roles: ['parent'],
    });
  }

  const studentsByWorkspaceUserId = new Map(
    (parentStudentsResult.data ?? []).map((student) => [student.id, student])
  );

  return {
    profile: {
      id: user.id,
      email: privateDetailsResult.data?.email ?? user.email ?? null,
      display_name:
        profileResult.data?.display_name ??
        privateDetailsResult.data?.full_name ??
        user.email ??
        null,
      avatar_url: profileResult.data?.avatar_url ?? null,
    },
    workspaces: Array.from(workspaceMap.values()).sort((a, b) =>
      (a.name ?? '').localeCompare(b.name ?? '')
    ),
    linkedStudents: parentLinks
      .map((link) => {
        const student = studentsByWorkspaceUserId.get(
          link.student_workspace_user_id
        );
        if (!student || !workspaceMap.has(link.ws_id)) return null;
        return {
          id: link.student_workspace_user_id,
          platform_user_id: link.student_platform_user_id,
          workspace_user_id: link.student_workspace_user_id,
          workspace_id: link.ws_id,
          name: toDisplayName(student),
          email: student.email ?? null,
          avatar_url: student.avatar_url ?? null,
        } satisfies TulearnStudentSummary;
      })
      .filter((student): student is TulearnStudentSummary => Boolean(student)),
  };
}
